lib/puppet_x/puppetlabs/dsc_type_helpers.rb

```diff
@@ -130,6 +130,35 @@ def self.validate_mof_type(mof_type, embeddedinstance_name, name, value)
           end
         end
       end
+
+      def self.should_add_reboot_relationship(resource, reboot_resource, pending_relationships)
+        return false if !reboot_resource
+
+        # edge exists in graph from previous resource, so don't add
+        return false if resource.catalog.relationship_graph.edge?(resource, reboot_resource) ||
+          # newly formed edges not yet in catalog include the edge, so don't add
+          pending_relationships.any? { |e| e.source == resource && e.target == reboot_resource }
+
+        true
+      end
+
+      # if an edge already exists from Reboot[dsc_reboot] to this resource
+      # Puppet will recognize the cyclic dependency automatically and fail with:
+      # Error: Failed to apply catalog: Found 1 dependency cycle:
+      # (Dsc_file[foo] => Reboot[dsc_reboot] => Dsc_file[foo])
+      def self.ensure_reboot_relationship(resource, pending_relationships)
+        reboot_resource = resource.catalog.resource(:reboot, 'dsc_reboot')
+
+        # do nothing if no Reboot[dsc_reboot] or already an edge from resource to it
+        if should_add_reboot_relationship(resource, reboot_resource, pending_relationships)
+          # otherwise build resource[name] => Reboot[dsc_reboot]
+          edge = Puppet::Relationship.new(resource, reboot_resource,
+              { :callback => :refresh, :event => :ALL_EVENTS })
+          pending_relationships << edge
+        end
+
+        pending_relationships
+      end
```

Comment:
Disregard, there are acceptance tests that catch this, so I'm good with that. It may be a little hard to test in specs so acceptance makes total sense to catch it.